### PR TITLE
Init harec and hare

### PR DESCRIPTION
**Pull request by ghost**
###### Description of changes

[Hare](harelang.org) is a systems programming language designed to be simple, stable, and robust. It has been in development for over a year and was publicly announced about two days ago. This commit packages `hare` and its bootstrap compiler `harec` for nixpkgs in accordance with Hare’s [instructions for distribution](https://harelang.org/distributions).

Two comments from me (as a maintainer). Firstly, I would greatly appreciate if one or more nixpkg maintainers with more experience packaging compilers felt excited enough to join in as maintainers as I am a little bit out of my comfort zone having only touched Julia in this domain in the past. Secondly, there is support for Linux aarch64 and riscv64, but I lack both experience and hardware to test that supporting these would work as intended. Thus, if anyone has the technical chops for this, please feel free to join and send a patch (or two!).

Comments are as always welcome, I am still learning Nix.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ninjin** [comment]
Not sure why aarch64-darwin is pending as it is not even supported in the first place. '^^

File: pkgs/development/compilers/hare/hare.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    sed -e "s|^PREFIX = .*|PREFIX = $out|" \
+        -e "s|^LOCALSRCDIR = .*|LOCALSRCDIR = $out/local/src/hare|" \
+        -i config.mk
+  '';
+
+  # Default cache location is user home, not writeable at Nix build time.
+  preBuild = ''
+    export HARECACHE="$NIX_BUILD_TOP"
+  '';
+  buildInputs = [ qbe harec scdoc makeWrapper ];
```
Comment (alyssais) [inline pkgs/development/compilers/hare/hare.nix:26]: At least `scdoc` and `makeWrapper` look to me like they should be native build inputs.  Try cross-compiling or setting `strictDeps = true;` (good practice anyway) to check.

File: pkgs/development/compilers/hare/hare.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    repo = pname;
+    rev = "a9c144a8d09cbdfee82857f926a413fea0e57298";
+    sha256 = "1z1ryqkg4vswzbnabw2nk7bdl1jjlz3jfqw8h3nylc5mr2zh3chx";
+  };
+
+  configurePhase = ''
+    cp config.example.mk config.mk
+    sed -e "s|^PREFIX = .*|PREFIX = $out|" \
+        -e "s|^LOCALSRCDIR = .*|LOCALSRCDIR = $out/local/src/hare|" \
+        -i config.mk
```
Comment (alyssais) [inline pkgs/development/compilers/hare/hare.nix:19]: Does setting these with `makeFlags` not work?

**Review by alyssais — CHANGES_REQUESTED** [review]
Good first attempt, thanks!

**Comment by alyssais** [comment]
> Not sure why aarch64-darwin is pending as it is not even supported in the first place. '^^

It's pending because the first thing it'll do is check whether it's supported, but it has to be started first.

**Review by ghost — COMMENTED** [review]
All comments incorporated and new version force pushed.

**Comment by Artturin** [comment]
a `passthru.tests.simple` which tries to compile something would be nice


that would also allow it to be tested on the platforms you don't use (by ofborg).

**Comment by Artturin** [comment]
:+1: 
adding a test would still be good

**Pull request closed without merging**